src/core/core.layouts.js

```diff
@@ -3,55 +3,178 @@
 var defaults = require('./core.defaults');
 var helpers = require('../helpers/index');
 
+var extend = helpers.extend;
+
 function filterByPosition(array, position) {
 	return helpers.where(array, function(v) {
-		return v.position === position;
+		return v.pos === position;
 	});
 }
 
 function sortByWeight(array, reverse) {
 	array.forEach(function(v, i) {
-		v._tmpIndex_ = i;
+		v.index = i;
+		v.weight = v.box.weight;
 		return v;
 	});
 	array.sort(function(a, b) {
 		var v0 = reverse ? b : a;
 		var v1 = reverse ? a : b;
 		return v0.weight === v1.weight ?
-			v0._tmpIndex_ - v1._tmpIndex_ :
+			v0.index - v1.index :
 			v0.weight - v1.weight;
 	});
-	array.forEach(function(v) {
-		delete v._tmpIndex_;
-	});
 }
 
-function findMaxPadding(boxes) {
-	var top = 0;
-	var left = 0;
-	var bottom = 0;
-	var right = 0;
-	helpers.each(boxes, function(box) {
-		if (box.getPadding) {
-			var boxPadding = box.getPadding();
-			top = Math.max(top, boxPadding.top);
-			left = Math.max(left, boxPadding.left);
-			bottom = Math.max(bottom, boxPadding.bottom);
-			right = Math.max(right, boxPadding.right);
-		}
-	});
+function wrapBoxes(boxes, availableWidth) {
+	var layoutBoxes = [];
+	var i, ilen, box, isHorizontal;
+
+	for (i = 0, ilen = (boxes || []).length; i < ilen; ++i) {
+		box = boxes[i];
+		isHorizontal = box.isHorizontal();
+		layoutBoxes.push({
+			pos: box.position,
+			box: box,
+			horizontal: isHorizontal,
+			// store width already to fullWidth horizontal boxes
+			width: isHorizontal && box.fullWidth && availableWidth
+		});
+	}
+	return layoutBoxes;
+}
+
+function buildLayoutBoxes(boxes, availableWidth) {
+	var layoutBoxes = wrapBoxes(boxes, availableWidth);
+	var left, right, top, bottom;
+
+	left = filterByPosition(layoutBoxes, 'left');
+	right = filterByPosition(layoutBoxes, 'right');
+	top = filterByPosition(layoutBoxes, 'top');
+	bottom = filterByPosition(layoutBoxes, 'bottom');
+
+	// Sort boxes by weight. A higher weight is further away from the chart area
+	sortByWeight(left, true);
+	sortByWeight(right);
+	sortByWeight(top, true);
+	sortByWeight(bottom);
+
 	return {
-		top: top,
-		left: left,
-		bottom: bottom,
-		right: right
+		lt: left.concat(top),
+		rb: right.concat(bottom),
+		area: filterByPosition(layoutBoxes, 'chartArea'),
+		v: left.concat(right),
+		h: top.concat(bottom)
 	};
 }
 
-function addSizeByPosition(boxes, size) {
-	helpers.each(boxes, function(box) {
-		size[box.position] += box.isHorizontal() ? box.height : box.width;
+function updatePadding(padding, boxes) {
```

Comment:
I think it'd be clearer to call the method `updateMaxPadding` and the first parameter `maxPadding`

Or better yet would be to remove the mutation:

```
function calculateMaxPadding(boxes) {
	var maxPadding = {};
	var boxPadding, i, ilen, box;

	for (i = 0, ilen = boxes.length; i < ilen; ++i) {
		box = boxes[i].box;
		if (box.getPadding) {
			boxPadding = box.getPadding();
			maxPadding.top = Math.max(maxPadding.top, boxPadding.top);
			maxPadding.left = Math.max(maxPadding.left, boxPadding.left);
			maxPadding.bottom = Math.max(maxPadding.bottom, boxPadding.bottom);
			maxPadding.right = Math.max(maxPadding.right, boxPadding.right);
		}
	}
	return maxPadding;
}
```

Reply:
renamed, but kept mutation because the object reference is kept in `update`.